File: Week_09/438/find_all_anagrams_in_a_string2.go
package main

func findAnagrams2(s string, p string) []int {
	res := make([]int, 0)

	counter := make([]int, 26)
	for i := 0; i < len(p); i++ {
		counter[p[i]-'a']++
	}
	tmpCounter := make([]int, 26)
	for left, right := 0, 0; right < len(s); right++ {
		tmpCounter[s[right]-'a']++
		for tmpCounter[s[right]-'a'] > counter[s[right]-'a'] {
			tmpCounter[s[left]-'a']--
			left++
		}
		if right-left+1 == len(p) {
			res = append(res, left)
		}
	}
	return res
}

File: Week_09/438/find_all_anagrams_in_a_string2.py
from typing import List


class Solution:
    def findAnagrams(self, s: str, p: str) -> List[int]:
        counter = [0] * 26
        for c in p:
            counter[ord(c) - ord('a')] += 1
        res = []

        left, right = 0, 0
        tmp_counter = [0] * 26
        while right < len(s):
            tmp_counter[ord(s[right]) - ord('a')] += 1
            while tmp_counter[ord(s[right]) - ord('a')] > counter[ord(s[right]) - ord('a')]:
                tmp_counter[ord(s[left]) - ord('a')] -= 1
                left += 1
            if right - left + 1 == len(p):
                res.append(left)
            right += 1
        return res
